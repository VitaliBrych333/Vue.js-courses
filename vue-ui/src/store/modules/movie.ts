import axios from 'axios'
import { Film, Films } from '@/components/interfaces/interfaces'
import { Module } from 'vuex'

const sortBy = (value: string, data: Array<Film>) => {
  let sortData: Array<Film> = []

  switch (value) {
    case 'vote_average':
      sortData = data.sort(
        (a: Film, b: Film) => b.vote_average - a.vote_average
      )
      break
    case 'release_date':
      sortData = data.sort(
        (a: Film, b: Film) =>
          new Date(b.release_date).valueOf() -
          new Date(a.release_date).valueOf()
      )
      break
    default:
      break
  }

  return sortData
}

const movie: Module<any, any> = {
  namespaced: true,

  state: {
    movies: { data: [], totalAmount: 0 },
    moviesByCriteria: { data: [], totalAmount: 0 } as Films,
    filmId: {},
    url: '',
    search: '',
    sortBy: 'release_date',
    genre: 'All',
    loading: false,
    error: null
  },

  mutations: {
    FETCH_MOVIES_BEGIN(state) {
      state.loading = true
      state.error = null
    },
    FETCH_MOVIES_SUCCESS(state, payload) {
      state.movies = payload.newMovies
      state.moviesByCriteria = payload.newMoviesByCriteria
    },
    FETCH_MOVIES_MORE(state, payload) {
      state.movies.data = [
        ...state.movies.data,
        ...(payload.newMovies.data as [])
      ]
      state.movies.totalAmount = payload.newMovies.totalAmount
      state.moviesByCriteria.data = [
        ...state.moviesByCriteria.data,
        ...(payload.newMoviesByCriteria.data as [])
      ]
      state.moviesByCriteria.totalAmount =
        payload.newMoviesByCriteria.totalAmount
    },
    FETCH_MOVIES_FAILURE(state, error) {
      state.movies = { data: [], totalAmount: 0 }
      state.loading = false
      state.error = error
    },
    FETCH_FILMID_BEGIN(state) {
      state.loading = true
      state.error = null
    },
    FETCH_FILMID_SUCCESS(state, filmId) {
      state.loading = false
      state.filmId = filmId
    },
    FETCH_FILMID_FAILURE(state, error) {
      state.loading = false
      state.error = error
      state.filmId = {}
    },
    SET_URL(state, value) {
      state.url = value
    },
    SET_SEARCH(state, value) {
      state.search = value
    },
    SORT(state, value) {
      let typeSort = ''

      switch (value) {
        case 'Release date':
          typeSort = 'release_date'
          break
        case 'Rating':
          typeSort = 'vote_average'
          break
        default:
          break
      }

      const { data } = state.moviesByCriteria

      if (data && data.length) {
        const sortData = sortBy(typeSort, data.slice())

        state.moviesByCriteria = {
          data: sortData,
          totalAmount: state.moviesByCriteria.totalAmount
        }
      }

      state.sortBy = typeSort
    },
    SET_GENRE(state, genre) {
      let totalCount

      let sortDataByGenre = sortBy(state.sortBy, state.movies.data.slice())

      if (genre !== 'All') {
        sortDataByGenre = sortDataByGenre
          .slice()
          .filter(film => film.genres.includes(genre))
        totalCount = sortDataByGenre.length
      } else {
        totalCount = state.movies.totalAmount
      }

      state.moviesByCriteria = {
        data: sortDataByGenre,
        totalAmount: totalCount
      }
      state.genre = genre
    }
  },

  actions: {
    FETCH_MOVIES({ commit }, payload) {
      const url = `movies?sortBy=${payload.sortBy}&sortOrder=desc&search=${payload.search}&searchBy=title`
      commit('SET_URL', url)
      commit('FETCH_MOVIES_BEGIN')

      return axios
        .get(`${url}&offset=${payload.offset}`)
        .then(res => {
          commit('FETCH_MOVIES_SUCCESS', {
            newMovies: res.data,
            newMoviesByCriteria: res.data
          })
          commit('SET_GENRE', 'All')
          commit('SET_SEARCH', payload.search)
        })
        .catch(err => commit('FETCH_MOVIES_FAILURE', err))
    },
    DOWNLOAD_MORE({ commit, state }, offset) {
      commit('FETCH_MOVIES_BEGIN')

      return axios
        .get(`${state.url}&offset=${offset}`)
        .then(res => {
          commit('FETCH_MOVIES_MORE', {
            newMovies: res.data,
            newMoviesByCriteria: res.data
          })
        })
        .catch(err => commit('FETCH_MOVIES_FAILURE', err))
    },
    DELETE_MOVIE({ commit }, payload) {
      return axios
        .delete(`movies/${payload.id}`)
        .then(() => {
          commit('FETCH_MOVIES_SUCCESS', {
            newMovies: payload.newMovies,
            newMoviesByCriteria: payload.newMoviesByCriteria
          })
          commit('windowMessage/SHOW_DELETE_PAGE', false, { root: true })
        })
        .catch(err => err)
    },
    ADD_MOVIE({ commit }, movie) {
      return axios
        .post('movies', JSON.stringify(movie))
        .then(() => {
          commit('windowMessage/SHOW_ADD_PAGE', false, { root: true })
        })
        .catch(err => err)
    },
    UPDATE_MOVIE({ commit, state }, payload) {
      return axios
        .put('movies', JSON.stringify(payload.movie))
        .then(() => {
          payload.movie.id === state.filmId.id
            ? commit('FETCH_FILMID_SUCCESS', payload.movie)
            : undefined
          commit('FETCH_MOVIES_SUCCESS', {
            newMovies: payload.newMovies,
            newMoviesByCriteria: payload.newMoviesByCriteria
          })
          commit('windowMessage/SHOW_EDIT_PAGE', false, { root: true })
        })
        .catch(err => err)
    },
    FETCH_MOVIES_BY_GENRE({ commit }, payload) {
      const url = `movies?sortBy=${payload.sortBy}&sortOrder=desc&searchBy=genres&filter=${payload.filterValue}`
      commit('FETCH_MOVIES_BEGIN')
      commit('SET_URL', url)
      return axios
        .get(`${url}&offset=0`)
        .then(res => {
          commit('FETCH_MOVIES_SUCCESS', {
            newMovies: res.data,
            newMoviesByCriteria: res.data
          })
          commit('SET_GENRE', 'All')
        })
        .catch(err => commit('FETCH_MOVIES_FAILURE', err))
    },
    FETCH_MOVIE_ID({ commit }, id) {
      commit('FETCH_FILMID_BEGIN')

      return axios
        .get(`movies/${id}`)
        .then(res => {
          commit('FETCH_FILMID_SUCCESS', res.data)
        })
        .catch(err => commit('FETCH_FILMID_FAILURE', err))
    }
  }
}

export default movie
