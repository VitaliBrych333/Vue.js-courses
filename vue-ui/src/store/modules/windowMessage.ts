import { Module } from 'vuex'

const windowMessage: Module<any, any> = {
  namespaced: true,

  state: {
    filmEdit: null,
    isShowEditPage: false,
    isShowDeletePage: false,
    isShowAddPage: false
  },

  mutations: {
    SET_EDIT_FILM(state, payload) {
      if (payload.actionType) {
        const showWindowEdit = payload.actionType === 'Edit'
        state.isShowEditPage = showWindowEdit
        state.isShowDeletePage = !showWindowEdit
      }
      state.filmEdit = payload.filmEdit
    },
    SHOW_EDIT_PAGE(state, value) {
      state.isShowEditPage = value
    },
    SHOW_DELETE_PAGE(state, value) {
      state.isShowDeletePage = value
    },
    SHOW_ADD_PAGE(state, value) {
      state.isShowAddPage = value
      state.filmEdit = null
    }
  }
}

export default windowMessage
