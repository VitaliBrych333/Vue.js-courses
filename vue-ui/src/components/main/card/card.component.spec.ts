import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import movie from '@/store/modules/movie'
import windowMessage from '@/store/modules/windowMessage'
import Card from './card.component.vue'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Card.vue', () => {
  let store: any
  const film = {
    id: 1,
    title: 'test',
    tagline: '',
    vote_average: 1,
    vote_count: 1,
    release_date: '2020-01-01',
    poster_path: 'http://',
    overview: 'test',
    budget: 1,
    revenue: 1,
    genres: [{ code: 'adventure', name: 'Adventure' }],
    runtime: 1
  }

  beforeEach(
    () =>
      (store = new Vuex.Store({
        state: {},
        modules: {
          movie,
          windowMessage
        }
      }))
  )

  it('should render correctly', () => {
    const wrapper = mount(Card, {
      store,
      localVue,
      propsData: {
        info: film
      }
    })
    const link = wrapper.findAll('a').at(1)
    const svg = wrapper.find('svg')
    const div = wrapper.findAll('div').at(0)

    div.trigger('mouseleave')
    svg.trigger('click')
    link.trigger('click')

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render correctly with props info', () => {
    const newFilm = Object.assign(film, { release_date: '' })
    const wrapper = mount(Card, {
      store,
      localVue,
      propsData: {
        info: newFilm,
        filmEdit: newFilm
      }
    })
    const link = wrapper.findAll('a').at(1)
    const svg = wrapper.find('svg')
    const div = wrapper.findAll('div').at(0)

    div.trigger('mouseleave')
    svg.trigger('mouseenter')
    link.trigger('click')

    expect(wrapper.props('info')).toEqual(newFilm)
  })
})
