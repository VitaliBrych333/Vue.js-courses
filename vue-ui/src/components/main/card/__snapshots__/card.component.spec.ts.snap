// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card.vue should render correctly 1`] = `
<div
  class="item"
>
  <!---->
   
  <div
    class="card mb-2"
  >
    <!---->
    <!---->
    <div
      class="card-body"
    >
      <!---->
      <!---->
      <svg
        aria-label="three dots vertical"
        class="bi-three-dots-vertical b-icon bi"
        fill="currentColor"
        focusable="false"
        height="1em"
        role="img"
        style="font-size: 200%; display: none;"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M9.5 13a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0zm0-5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0zm0-5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0z"
            fill-rule="evenodd"
          />
        </g>
      </svg>
       
      <div
        class="img-wrapper"
      >
        <img
          alt="Image"
          src="http://"
        />
      </div>
       
      <div>
        <a
          class="title"
          href="/movies/1"
          target="_self"
        >
          test
        </a>
         
        <span
          class="badge badge-secondary"
        >
          2020
        </span>
      </div>
       
      <a
        class="genre"
        href="#"
        target="_self"
      >
        {
  "code": "adventure",
  "name": "Adventure"
}
      </a>
    </div>
    <!---->
    <!---->
  </div>
</div>
`;
